### PR TITLE
fix: correct LoRA initialization and forward pass under tensor parallelism

### PR DESCRIPTION
**Pull request by chen2021673**
## Summary

Fix LoRA loss divergence under tensor/data parallel training by making LoRA tensor-parallel initialization deterministic and aligning the forward collective order between base linear and LoRA linear paths.

This PR changes TP LoRA parallel linear modules to compute the base shard and LoRA shard locally first, add them before communication, and then run a single TP collective on the combined output. It also adds rank-aware `Broadcast`/`Scatter` ProcessGroup APIs and updates LoRA weight loading to support loading full saved LoRA tensors into TP-sharded model parameters.

It also adds rank-aware `Broadcast`/`Scatter` ProcessGroup APIs, updates LoRA adapter save/load to use portable full tensors under TP, handles packed QKV LoRA tensors, and keeps adapter IO compatible with DDP-wrapped models.

## Motivation

In tensor-parallel LoRA training, the base linear output and the LoRA update together form one logical linear output. Previously, the base path and LoRA path could run separate TP collectives and add their outputs afterward:

```cpp
base = TPCollective(base_i);
lora = TPCollective(lora_i);
out = base + lora;
```

This changes the floating-point communication/reduction order compared with computing the logical local contribution first:

```cpp
out_i = base_i + lora_i;
out = TPCollective(out_i);
```

The separate-collective path can introduce numerical divergence across TP/DDP configurations.

Replicated or sharded LoRA parameters need rank-consistent initialization so every TP rank starts from the same logical LoRA weights.

### LoRA parallel linear forward

* Update `LoRAColumnParallelLinear` to:

  * compute the base shard locally;
  * compute the LoRA shard locally;
  * add `base_shard + scaled_lora_shard` before TP gather;
  * run a single gather when `gather_output_` is enabled.

* Update `LoRARowParallelLinear` to:

  * compute the base shard locally;
  * compute the LoRA shard locally;
  * add `base_shard + scaled_lora_shard` before TP reduce/reduce-scatter;
  * apply bias after the collective, matching the base `RowParallelLinear` behavior.

This makes the LoRA path follow the same logical communication boundary as the base parallel linear layer and avoids running separate collectives for base and LoRA outputs.

### Deterministic TP LoRA initialization

* Make replicated ColumnParallel `lora_A` consistent across TP ranks.
* Initialize the logical RowParallel `lora_A` from TP rank 0 and distribute the correct local shards to each TP rank.
* Ensure TP ranks observe the same logical LoRA initialization instead of independently sampling incompatible parameter values.

### ProcessGroup communication APIs

* Add rank-aware communication APIs:

  * `ProcessGroup::Broadcast(tensors, root_rank_in_group, async_op)`
  * `ProcessGroup::Scatter(output_tensors, input_tensors, root_rank_in_group, async_op)`
* Rename the old APIs to `BroadCast_` / `Scatter_` to avoid semantic ambiguity.
* Update existing autograd communication wrappers to call the legacy renamed APIs where appropriate.

### LoRA weight save/load

* Extend `LoadLoRAWeights` to support loading full saved LoRA weights into TP-sharded model parameters.
* When a destination parameter is TP-sharded, slice the loaded full tensor according to the current `tp_rank` before copying it into the local parameter.
* Update `SaveLoRAWeights` to export portable full LoRA weights under TP:

  * TP-sharded LoRA tensors are gathered back into full/unsharded tensors before writing.
  * All TP ranks in the primary DP/PP group participate in the save-side gather.
  * Only the rank with `dp == 0 && tp == 0 && pp == 0` writes the weight file.

### Packed QKV LoRA tensor handling

  * Handle attention `c_attn.lora_B` specially when it uses the packed QKV layout `[Q | K | V]`.
  * On save, restore full packed QKV tensors from per-rank `[Qi | Ki | Vi]` shards.
  * On load, split full packed QKV tensors back into local `[Qi | Ki | Vi]` shards for the current TP rank.
  * Add helper utilities for packed QKV slicing/restoring and unit tests covering GPT-style and GQA-style QKV layouts.
 
### DDP adapter state dict compatibility

  * Expose the wrapped module from `DistributedDataParallel` and save LoRA adapters from the inner model.
  * This matches the common `model.module.state_dict()` behavior for DDP-wrapped models.
  * Consume a legacy leading `module.` prefix when loading LoRA adapter state dicts.
  * Add test coverage for DDP `module.` prefix compatibility.
  
## Test

The loss values in the LoRA tests fluctuate slightly, which is expected because the tests now use saved fixed initialization values.

image

The performance variance is concentrated in the `lora/bfloat16/distopt` tests and is treated as normal fluctuation.

image

### CODE REVIEW

File: infini_train/src/nn/lora/lora_utils.cc [hunk 1]
```diff
@@ -392,10 +393,30 @@ void LoadLoRAWeights(std::shared_ptr<Module> model, const std::string &filepath)
...
+                        break;
+                    }
+                }
+                CHECK(shard_dim >= 0) << "LoadLoRAWeights: shape mismatch for " << name
+                                      << " but no differing dim found";
+                int tp_size = parallel::global::GetTensorParallelSize();
+                int64_t shard_size = dims[shard_dim] / tp_size;
+                int64_t start = parallel::tp_rank * shard_size;
+                auto sliced = cpu_tensor->Slice(shard_dim, start, start + shard_size);
+                dst->CopyFrom(sliced);
```
Comment (Chamberlain0w0) [inline infini_train/src/nn/lora/lora_utils.cc]: 现在模型实现中，Attention 里面第一个 Linear 把 QKV 合并了（实际排列上是 [Q | K | V]），LoadFromLLMC 里面有对应的处理（如 TP=4，则切成 [Q0 Q1 Q2 Q3 | K0 K1 K2 K3 | V0 V1 V2 V3]，然后依次拼接 [Qi | Ki | Vi] load 到每个 tp rank 上）。所以这块的切分逻辑对于这种情况没法适用，得想想看怎么能特别处理一下

File: infini_train/src/nn/lora/lora_parallel_linear.cc [hunk 2]
```diff
@@ -89,22 +90,34 @@ LoRAColumnParallelLinear::LoRAColumnParallelLinear(std::shared_ptr<parallel::Col
...
+
+    if (parallel::global::GetTensorParallelSize() > 1) {
+        const auto global_rank = device_.Rank().GlobalRank();
+        auto *tp_group = parallel::ProcessGroupFactory::Instance(device_.type())
+                             ->Get(parallel::GetTensorParallelProcessGroupName(global_rank));
+        const int tp_rank = tp_group->GetGroupRank(global_rank);
+
+        if (tp_rank == 0) {
+            if (config_.use_kaiming_a) {
+                init::KaimingUniform(parameters_[kParamLoraAName], config_.kaiming_a_param);
```
Comment (Chamberlain0w0) [inline infini_train/src/nn/lora/lora_parallel_linear.cc:111]: 在这里做各个 tp rank 各自 init，可能会导致 DP 组之间无法达到权重一致的结果。现在的情况可能会因为都用默认 seed 导致恰好能对上，但是从代码看，语义上还是各初始化各的，DP 组之间有可能权重对不齐。原则上应该保证 DDP model 最后再做一步参数的广播/复制，从而确保在实际执行 forward 前各个能对应上的 dp rank 上面模型参数是相同的，但这里我知道改的话会不会影响其他 lora 基建；最简单的就是 `tp_group->Broadcast({parameters_[kParamLoraAName]}, 0);`  后面再加个 dp_group 的 broadcast。

File: infini_train/src/nn/lora/lora_utils.cc
Comment (Chamberlain0w0) [inline infini_train/src/nn/lora/lora_utils.cc:615]: save 似乎 save 的是 tp local shard tensor，跟 load 逻辑对不上了？load 应该默认读 full tensor 吧？

**Review by Chamberlain0w0 — CHANGES_REQUESTED** [review]

File: infini_train/src/nn/lora/lora_utils.cc (hunk 1 above)
Comment (chen2021673) [reply infini_train/src/nn/lora/lora_utils.cc]: 加一个特判，在 LoadLoRAWeights 里识别 attn.c_attn 这种输出维分片参数，把 full [Q|K|V] 切成 local [Qi|Ki|Vi]。

File: infini_train/src/nn/lora/lora_utils.cc
Comment (chen2021673) [reply infini_train/src/nn/lora/lora_utils.cc:615]: 修改：SaveLoRAWeights 默认写 full tensor，LoadLoRAWeights 也默认读 full tensor，再按当前 TP rank 切片加载

File: infini_train/src/nn/lora/lora_utils.cc (hunk 1 above)
Comment (chen2021673) [reply infini_train/src/nn/lora/lora_utils.cc]: checkpoint 没单独处理这个类型，load 的时候必须对得上 TP 的配置。但这里因为要和单机对齐初始化参数，有这个功能需求。

File: infini_train/src/nn/lora/lora_parallel_linear.cc (hunk 2 above)
Comment (chen2021673) [reply infini_train/src/nn/lora/lora_parallel_linear.cc:111]: 这里我觉得应该写成 DDP 构造后的通用参数同步，而不是 LoRA 专用补丁。可以单独提PR。

**Review by Chamberlain0w0 — APPROVED** [review]

File: infini_train/src/nn/lora/lora_parallel_linear.cc (hunk 2 above)
Comment (kilinchange) [reply infini_train/src/nn/lora/lora_parallel_linear.cc:111]: 这里留个 FIXME 记一下吧。

File: tests/lora/test_lora.cc [hunk 3]
```diff
@@ -30,6 +31,54 @@ static float TensorSum(const std::shared_ptr<Tensor> &t) {
...
+    if (device.IsCPU()) {
+        return cpu_tensor;
+    }
+
+    auto device_tensor = std::make_shared<Tensor>(std::vector<int64_t>{rows, cols}, DataType::kFLOAT32, device);
+    device_tensor->CopyFrom(cpu_tensor);
+    return device_tensor;
+}
+
+static std::shared_ptr<Tensor> ToCpuTensor(const std::shared_ptr<Tensor> &tensor) {
```
Comment (kilinchange) [inline tests/lora/test_lora.cc]: 没必要封装这个函数吧，直接调用 Tensor 的 To 方法就行

File: scripts/run_models_and_profile.bash [hunk 4]
```diff
@@ -259,8 +259,9 @@ run_and_log() {
         > "$log_path"
     fi
 
-    # Write the current run command to the log
-    echo "[COMMAND] $cmd" >> "$log_path"
+    # Write the current run command to the log (expand $LORA_WEIGHTS_DIR)
+    local expanded_cmd="${cmd//\$LORA_WEIGHTS_DIR/$LORA_WEIGHTS_DIR}"
+    echo "[COMMAND] $expanded_cmd" >> "$log_path"
```
Comment (kilinchange) [inline scripts/run_models_and_profile.bash:313]: 这里是不是没必要改，expanded_cmd 等价于没有做替换，和原来是一样的。

File: infini_train/src/nn/lora/lora_utils.cc [hunk 5]
```diff
@@ -11,15 +12,351 @@
...
+    return shardings;
+}
+
+LoRATensorSharding GetLoRATensorSharding(const std::unordered_map<std::string, LoRATensorSharding> &shardings,
+                                         const std::string &name) {
+    auto it = shardings.find(name);
+    return it == shardings.end() ? LoRATensorSharding::kReplicated : it->second;
+}
+
+std::shared_ptr<Tensor> GatherTensorParallelShard(const std::shared_ptr<Tensor> &tensor, int64_t dim) {
```
Comment (kilinchange) [inline infini_train/src/nn/lora/lora_utils.cc]: 这个函数与 infini_train/src/nn/parallel/tensor_parallel.cc 里匿名函数 GatherAlongFirstDim/GatherAlongLastDim 有所重复，建议复用，有必要的话可以抽出公共接口放入 infini_train/include/nn/parallel/utils.h。具体修改细节可以跟 @Chamberlain0w0 讨论下。

File: infini_train/src/nn/lora/lora_utils.cc [hunk 6]
```diff
@@ -11,15 +12,351 @@
...
+    auto sliced = src->Slice(shard_dim, start, start + shard_size);
+    CHECK(sliced->Dims() == dst_dims) << "LoadLoRATensorIntoModel: shard shape mismatch for " << name;
+    dst->CopyFrom(sliced);
+}
+
+} // namespace
+
+namespace detail {
+
+std::shared_ptr<Tensor> SlicePackedQKVRowsForTensorParallel(const std::shared_ptr<Tensor> &full_tensor, int64_t q_rows,
```
Comment (kilinchange) [inline infini_train/src/nn/lora/lora_utils.cc:255]: 这部分逻辑理论上 tp 应该复用，但是由于当前 tp 是基于 offset 读输入权重文件，不是从已有 Tensor 切分，所以暂时没法直接复用，先留个 TODO，等之后 ckpt 基建稳定后再考虑如何统一。

**Review by kilinchange — CHANGES_REQUESTED** [review]

File: scripts/run_models_and_profile.bash (hunk 4 above)
Comment (chen2021673) [reply scripts/run_models_and_profile.bash:313]: 这里是把字面文本 $LORA_WEIGHTS_DIR 替换成真正的值

File: tests/lora/test_lora.cc (hunk 3 above)
Comment (chen2021673) [reply tests/lora/test_lora.cc]: done

File: infini_train/src/nn/lora/lora_utils.cc (hunk 5 above)
Comment (chen2021673) [reply infini_train/src/nn/lora/lora_utils.cc]: done

File: infini_train/src/nn/lora/lora_utils.cc (hunk 6 above)
Comment (chen2021673) [reply infini_train/src/nn/lora/lora_utils.cc:255]: done

File: infini_train/src/nn/lora/lora_parallel_linear.cc (hunk 2 above)
Comment (chen2021673) [reply infini_train/src/nn/lora/lora_parallel_linear.cc:111]: done

**Review by kilinchange — APPROVED** [review]

**Comment by kilinchange** [comment]
麻烦修改一下 3b6088ed5317ee2c579caa1e6ae832259bad5015 这个 commit 的 message，然后再跑一遍全量测试吧。

**Review by Chamberlain0w0 — APPROVED** [review]

**Comment by chen2021673** [comment]
> 麻烦修改一下 [3b6088e](https://github.com/InfiniTensor/InfiniTrain/commit/3b6088ed5317ee2c579caa1e6ae832259bad5015) 这个 commit 的 message，然后再跑一遍全量测试吧。

已修改。
测试截图已更新到PR描述。
重复实验，loss 能够对齐。
image
